rgw/sts: correcting authentication in case s3 ops are directed to a primary from secondary after assumerole.

pritha-srivastava opened the pull request:
rgw/sts: by-passing authentication using temp creds
in case the request is forwarded from secondary in a multi-site setup. authenticating with the system user creds of which are used to sign the request.
Permissions are still derived from the role.





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

pritha-srivastava commented:
@cbodley , @mattbenjamin @smanjara : requesting a quick feedback on the changes in this PR. Doc changes and teuthology changes still to be done.

pritha-srivastava commented:
The changes have been tested locally and work fine.

pritha-srivastava commented:
The test case also works fine locally. I want to schedule a teuthology run, where do I add rgw config options for multisite tests? @smanjara @cbodley @alimaredia

smanjara commented:
> The test case also works fine locally. I want to schedule a teuthology run, where do I add rgw config options for multisite tests? @smanjara @cbodley @alimaredia

hi @pritha-srivastava you could add them in [qa/suites/rgw/multisite/overrides.yaml](https://github.com/ceph/ceph/blob/main/qa/suites/rgw/multisite/overrides.yaml)

pritha-srivastava commented:
> > The test case also works fine locally. I want to schedule a teuthology run, where do I add rgw config options for multisite tests? @smanjara @cbodley @alimaredia
> 
> hi @pritha-srivastava you could add them in [qa/suites/rgw/multisite/overrides.yaml](https://github.com/ceph/ceph/blob/main/qa/suites/rgw/multisite/overrides.yaml)

Thank you @smanjara

smanjara commented:
jenkins test make check

mattbenjamin commented:
@pritha-srivastava @smanjara what is this PR stalled on?

pritha-srivastava commented:
> @pritha-srivastava @smanjara what is this PR stalled on?

I will rebase and update the PR.

pritha-srivastava commented:
test scheduled here: https://pulpito.ceph.com/prsrivas-2024-08-14_05:29:13-rgw:multisite-wip-rgw-assume-role-multisite-distro-default-smithi/

pritha-srivastava commented:
jenkins make check

pritha-srivastava commented:
jenkins test make check

smanjara commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -6492,10 +6493,36 @@ rgw::auth::s3::STSEngine::authenticate(
   //Authentication
+  std::string secret_access_key;
   //Check if access key is not the same passed in by client
   if (token.access_key_id != _access_key_id) {
```
this condition check probably ensures that the system user credentials is used to sign the request _only_ when the zone is secondary? otherwise we can check for` if (!driver->is_meta_master())`

pritha-srivastava replied:
the request is forwarded from the secondary to the primary so the check would be to see if the current node is the master which is if (driver->is_meta_master()) ?

smanjara replied:
it looks like you are already checking for `user->get_info().system` to ensure that authenticating against session token credentials is by-passed only for secondary and not for requests originating from primary. that should be enough.

pritha-srivastava replied:
ok

pritha-srivastava replied:
does user->get_info().system guarantee that the code operates only on the primary? Or to be on the safer side, do you think I must add the check for driver->is_meta_master()

smanjara commented:
> > @pritha-srivastava @smanjara what is this PR stalled on?
> 
> I will rebase and update the PR.

sorry, just reviewing this now.

adamemerson commented:
jenkins test api

adamemerson commented:
@pritha-srivastava Anything I can do to help unblock this? If it's complete and approved and you don't object I could try rebasing it so it can go through QA.

pritha-srivastava commented:
@adamemerson : I will rebase it today. I am unable to schedule a teuthology run as my sepia lab access is being sorted as of now, so I will need help testing it

pritha-srivastava commented:
> @adamemerson : I will rebase it today. I am unable to schedule a teuthology run as my sepia lab access is being sorted as of now, so I will need help testing it

@adamemerson @smanjara : I need help rebasing the PR wrt the test case I added, the idea was to add an alt_user credentials to the zone conn and use it for sts connection. Now there are non_account_alt_user and non_account_zone_conn also, so should I still add the alt user credentials to zone conn or should I use non_account_zone_conn for the test case?

pritha-srivastava commented:
> > @adamemerson : I will rebase it today. I am unable to schedule a teuthology run as my sepia lab access is being sorted as of now, so I will need help testing it
> 
> @adamemerson @smanjara : I need help rebasing the PR wrt the test case I added, the idea was to add an alt_user credentials to the zone conn and use it for sts connection. Now there are non_account_alt_user and non_account_zone_conn also, so should I still add the alt user credentials to zone conn or should I use non_account_zone_conn for the test case?

I will remove alt_user that I added and re-use the non_account_alt_user that is already there and pass the same to Conn class and use it for the sts connection. Let me know if this sounds correct.

pritha-srivastava commented:
> > > @adamemerson : I will rebase it today. I am unable to schedule a teuthology run as my sepia lab access is being sorted as of now, so I will need help testing it
> > 
> > 
> > @adamemerson @smanjara : I need help rebasing the PR wrt the test case I added, the idea was to add an alt_user credentials to the zone conn and use it for sts connection. Now there are non_account_alt_user and non_account_zone_conn also, so should I still add the alt user credentials to zone conn or should I use non_account_zone_conn for the test case?
> 
> I will remove alt_user that I added and re-use the non_account_alt_user that is already there and pass the same to Conn class and use it for the sts connection. Let me know if this sounds correct.

actually I added alt_user as another user under an account, and rebased the PR, I need help in testing it now

pritha-srivastava commented:
@cbodley @smanjara : I need help with the test case that I added to the PR here, I want to add an alternate user that can be used to invoke AssumeRole. I had written the test prior to user accounts getting merged and now I am not getting it correct. from teuthology:
```55f208b2b/qa/../src/test/rgw/rgw_multi/tests.py", line 619, in test_bucket_create
2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:    zonegroup_conns = ZonegroupConns(zonegroup)
2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/github.com_pritha-srivastava_ceph_d840368ce38ace79b7642b1228ab90755f208b2b/qa/../src/test/rgw/rgw_multi/tests.py", line 507, in __init__
2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:    non_account_zone_conn = z.get_conn(non_account_user.credentials)
2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:TypeError: RadosZone.get_conn() missing 1 required positional argument: 'alt_user_credentials'```

cbodley commented:
> @cbodley @smanjara : I need help with the test case that I added to the PR here, I want to add an alternate user that can be used to invoke AssumeRole. I had written the test prior to user accounts getting merged and now I am not getting it correct. from teuthology:
> 
> ```
> 2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:    zonegroup_conns = ZonegroupConns(zonegroup)
> 2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/github.com_pritha-srivastava_ceph_d840368ce38ace79b7642b1228ab90755f208b2b/qa/../src/test/rgw/rgw_multi/tests.py", line 507, in __init__
> 2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:    non_account_zone_conn = z.get_conn(non_account_user.credentials)
> 2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:TypeError: RadosZone.get_conn() missing 1 required positional argument: 'alt_user_credentials'
> ```

Shilpa recently added a `test_bucket_create_with_tenant()` in https://github.com/ceph/ceph/pull/61433 that needed a tenanted user. we found that it was simplest for the test case to create its own user with `radosgw-admin user create`

pritha-srivastava commented:
> > @cbodley @smanjara : I need help with the test case that I added to the PR here, I want to add an alternate user that can be used to invoke AssumeRole. I had written the test prior to user accounts getting merged and now I am not getting it correct. from teuthology:
> > ```
> > 2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:    zonegroup_conns = ZonegroupConns(zonegroup)
> > 2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/github.com_pritha-srivastava_ceph_d840368ce38ace79b7642b1228ab90755f208b2b/qa/../src/test/rgw/rgw_multi/tests.py", line 507, in __init__
> > 2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:    non_account_zone_conn = z.get_conn(non_account_user.credentials)
> > 2025-04-14T09:49:36.016 INFO:tasks.rgw_multisite_tests:TypeError: RadosZone.get_conn() missing 1 required positional argument: 'alt_user_credentials'
> > ```
> 
> Shilpa recently added a `test_bucket_create_with_tenant()` in #61433 that needed a tenanted user. we found that it was simplest for the test case to create its own user with `radosgw-admin user create`

I looked at her test case, the difference in my test case is that the user credentials is needed for creating an sts connection. so the user has to be created elsewhere. And I need the alternate user to be present in the same zone as the other user.

cbodley commented:
> I looked at her test case, the difference in my test case is that the user credentials is needed for creating an sts connection. so the user has to be created elsewhere. And I need the alternate user to be present in the same zone as the other user.

i think the unnecessary "zone conn" abstraction is what's making this so difficult. that abstraction assumes there's only ever one identity, which isn't good enough for the test coverage we need

in the [test_bucket_create_with_tenant()](https://github.com/ceph/ceph/blob/2bd32f7/src/test/rgw/rgw_multi/tests.py#L618-L647) example, it's running `radosgw-admin user create` and constructing its own boto connection with those credentials

could you do similar with the sts and iam connections, instead of relying on changes to zone_rados.py? you can still use the `get_gateway_sts_connection()` helper function there, as long as it doesn't cache the result in `gateway.sts_connection`

> Can non_account_alt_user be used for this purpose? but I see that there are different zone connections for non_acct_user and non_account_alt_user

i'm not familiar with those, but cc @clwluvw who added them

clwluvw commented on src/test/rgw/rgw_multi/tests.py:
```diff
@@ -495,7 +502,8 @@ def __init__(self, zonegroup):
+            log.debug('ZonegroupConns::__init__ alt_user=%s', alt_user.name)
+            zone_conn = z.get_conn(user.credentials, alt_user.credentials)
             non_account_zone_conn = z.get_conn(non_account_user.credentials)
             non_account_alt_zone_conn = z.get_conn(non_account_alt_user.credentials)
```
With the changes to `get_conn()`, the current usage of `non_account_*` calls will be the issue.
I'm not sure about the motivation for passing alternate credentials to `ZoneConn`—it appears they are primarily used for creating the STS client. However, if we can extend `ZoneConn` to include support for an alternate user context, it might be possible to unify those independent `non_account_*` calls under the same framework.

clwluvw commented on src/rgw/rgw_auth.cc:
```diff
@@ -1304,6 +1304,9 @@ void rgw::auth::RoleApplier::modify_request_state(const DoutPrefixProvider *dpp,
     s->token_claims.emplace_back(it);
   }
+  if (is_system_request) {
+    s->system_request = true;
```
As [Casey mentioned](https://github.com/ceph/ceph/pull/61962#issuecomment-2819084605), I just wanted to check whether the idea introduced here—adding `is_impersonating` to `SysReqApplier`—might also be applicable in this context:
https://github.com/ceph/ceph/pull/61962/commits/bb2cf19e86304408221041b2483bfafc7b0a4445

The idea is that if `SysReqApplier` has `is_impersonating` set to true, it would automatically set `s->system_request = true` within `modify_request_state()`. Additionally, a helper function like `is_admin()` in the identity could be used to determine whether the request was impersonated for another user or a normal system request without any impersonation.

cbodley replied:
right. i think this is a minor conflict, so don't want to block either PR on it. we'll just have to follow up once one or the other merges

pritha-srivastava commented:
@cbodley @smanjara : the teuthology results are here: the tests are passing in both the jobs:

https://pulpito.ceph.com/prsrivas-2025-04-25_12:09:31-rgw:multisite-wip-rgw-assume-role-multisite-distro-default-smithi/
```2025-04-25T13:01:19.581 INFO:tasks.rgw_multisite_tests:rgw_multi.tests.test_assume_role_after_sync ... ok
```2025-04-25T13:03:06.512 INFO:tasks.rgw_multisite_tests:rgw_multi.tests.test_assume_role_after_sync ... ok
```

cbodley reviewed (approved):
approved regardless, lets get this through the full suite :rocket:

pritha-srivastava commented:
> alt_user_creds = gen_credentials()

yes, scheduled one just now

pritha-srivastava commented on src/test/rgw/test_multi.py:
```diff
@@ -254,6 +254,11 @@ def init(parse_args):
     non_account_alt_user_creds = gen_credentials()
     non_account_alt_user = multisite.User('nonaccountalttester', tenant=args.tenant)
 
+    alt_user_creds = gen_credentials()
+    log.debug('created alt_user_creds access key=%s secret=%s', alt_user_creds.access_key, alt_user_creds.secret)
+    alt_user = multisite.User('alt_tester', tenant=args.tenant, account='RGW11111111111111111')
+    log.debug('created alt_user=%s', alt_user.name)
```
I have some code cleanup to do, will do so and update the PR

pritha-srivastava commented:
Teuthology tests have passed, except two: https://pulpito.ceph.com/prsrivas-2025-04-27_07:35:37-rgw-wip-rgw-assume-role-multisite-distro-default-smithi/
1. d4n suite failure (which is known) - 
```
Command failed (workunit test rgw/run-d4n.sh) on smithi033 with status 1: 'mkdir -p -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && cd -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && CEPH_CLI_TEST_DUP_COMMAND=1 CEPH_REF=855db87f4addec8576708d56b6f6d6554caf8b37 TESTDIR="/home/ubuntu/cephtest" CEPH_ARGS="--cluster ceph" CEPH_ID="0" PATH=$PATH:/usr/sbin CEPH_BASE=/home/ubuntu/cephtest/clone.client.0 CEPH_ROOT=/home/ubuntu/cephtest/clone.client.0 CEPH_MNT=/home/ubuntu/cephtest/mnt.0 adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 3h /home/ubuntu/cephtest/clone.client.0/qa/workunits/rgw/run-d4n.sh'
```
3. multisite failure - 
```
FAIL: rgw_multi.tests.test_bucket_log_trim_after_delete_bucket_secondary_reshard
2025-04-27T11:18:26.234 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2025-04-27T11:18:26.234 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2025-04-27T11:18:26.235 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_teuthology_dc15ac4a651ab5968cb4ffaa9ef0ff1a02484ea2/virtualenv/lib/python3.10/site-packages/nose/case.py", line 170, in runTest
2025-04-27T11:18:26.235 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2025-04-27T11:18:26.235 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/github.com_pritha-srivastava_ceph_855db87f4addec8576708d56b6f6d6554caf8b37/qa/../src/test/rgw/rgw_multi/tests.py", line 1897, in test_bucket_log_trim_after_delete_bucket_secondary_reshard
2025-04-27T11:18:26.235 INFO:tasks.rgw_multisite_tests:    assert check_bucket_instance_metadata(zone.zone, test_bucket.name)
2025-04-27T11:18:26.235 INFO:tasks.rgw_multisite_tests:AssertionError
```
In the same run, the test in this PR has passed:
```
2025-04-27T11:17:17.370 INFO:tasks.rgw_multisite_tests:rgw_multi.tests.test_assume_role_after_sync ... ok
```

pritha-srivastava commented:
hence this PR is ready to be merged @cbodley @smanjara

cbodley commented:
does this not need backports? with account support for roles in squid, i think it would be nice to have there at least

smanjara commented:
> does this not need backports? with account support for roles in squid, i think it would be nice to have there at least

@pritha-srivastava I created a tracker https://tracker.ceph.com/issues/71113 for squid. would you please do the backport?

cbodley commented:
> @pritha-srivastava I created a tracker https://tracker.ceph.com/issues/71113 for squid. would you please do the backport?

thanks @smanjara. i added 'tentacle' to https://tracker.ceph.com/issues/71112 as well, since we forked on friday before this merged
